packages/core/contracts/oracle/implementation/StakerSnapshot.sol

```diff
@@ -0,0 +1,242 @@
+     *           STAKING TRACKERS           *
+     ****************************************/
+
+    uint256 public emissionRate;
+    uint256 public cumulativeStaked;
+    uint256 public rewardPerTokenStored;
+    uint256 public lastUpdateTime;
+
+    uint256 unstakeCoolDown;
+
+    struct VoterStake {
+        uint256 cumulativeStaked;
+        uint256 rewardsPaidPerToken;
+        uint256 outstandingRewards;
+        uint256 unstakeTime;
+        uint256 requestUnstake;
+        uint256 lastRequestIndexConsidered;
+        Snapshots snapshots;
+    }
+
+    mapping(address => VoterStake) public voterStakes;
+
+    struct Snapshots {
+        uint256[] ids;
+        uint256[] values;
+    }
+
+    Snapshots private _cumulativeStakedSnapshots;
+
+    // Snapshot ids increase monotonically, with the first value being 1. An id of 0 is invalid.
+    Counters.Counter private _currentSnapshotId;
+
+    // Reference to the voting token.
+    VotingToken public votingToken;
+
+    event Snapshot(uint256 id);
+
+    constructor(
+        uint256 _emissionRate,
+        uint256 _unstakeCoolDown,
+        address _votingToken,
+        address _timerAddress
+    ) Testable(_timerAddress) {
+        emissionRate = _emissionRate;
+        unstakeCoolDown = _unstakeCoolDown;
+        votingToken = VotingToken(_votingToken);
+    }
+
+    // Pulls tokens from users wallet and stakes them.
+    function stake(uint256 amount) public {
+        _updateTrackers(msg.sender);
+        updateSnapshot(msg.sender);
+
+        voterStakes[msg.sender].cumulativeStaked += amount;
+        cumulativeStaked += amount;
+
+        votingToken.transferFrom(msg.sender, address(this), amount);
+    }
+
+    function updateSnapshot(address voterAddress) public {
+        _updateAccountSnapshot(voterAddress);
+        _updateTotalStakedSnapshot();
+    }
+
+    function requestUnstake(uint256 amount) public {
+        _updateTrackers(msg.sender);
+        // Staker signals that they want to unstake. After signalling, their total voting balance is decreased by the
+        // signaled amount. This amount is not vulnerable to being slashed but also does not accumulate rewards.
+
+        require(voterStakes[msg.sender].cumulativeStaked >= amount, "Bad request amount");
+
+        voterStakes[msg.sender].requestUnstake = amount;
+        voterStakes[msg.sender].unstakeTime = getCurrentTime() + unstakeCoolDown;
+    }
+
+    // Note there is no way to cancel your unstake; you must wait until after unstakeTime and re-stake.
+
+    // If: a staker requested an unstake and time > unstakeTime then send funds to staker. Note that this method assumes
+    // that the `updateTrackers()
+    function executeUnstake() public {
+        _updateTrackers(msg.sender);
+        updateSnapshot(msg.sender);
+        VoterStake storage voterStake = voterStakes[msg.sender];
+        require(voterStake.unstakeTime != 0 && getCurrentTime() >= voterStake.unstakeTime, "Unstake time not passed");
```

Comment:
Similar comment to last time: https://github.com/UMAprotocol/protocol/pull/3971#discussion_r898374036. Is there any way to avoid this unstaking UX -- it seems like this will hurt the overall voting UX to have a delay like this.

In general, it seems sufficient to only have voters get rewards on the balance they had during the previous vote.